File: app/src/main/java/com/example/features/ViewPresenter.java
package com.example.features;

import android.support.annotation.NonNull;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

import static com.example.toolkit.Preconditions.checkNotNull;

public abstract class ViewPresenter<V extends PresenterView> {
  private final CompositeDisposable mCompositeSubscription = new CompositeDisposable();
  private V mView;

  protected ViewPresenter(@NonNull V view) {
    checkNotNull(view, "view cannot be null.");
    mView = view;
  }

  public void bind(@NonNull V view) {
    checkNotNull(view, "view cannot be null.");
    mView = view;
  }

  public void release() {
    clearSubscriptions();
    mView = null;
  }

  protected V getView() {
    return mView;
  }

  protected void addSubscription(Disposable disposable) {
    mCompositeSubscription.add(disposable);
  }

  public void clearSubscriptions() {
    mCompositeSubscription.clear();
  }
}

File: libs/toolkit-android/src/main/kotlin/br/com/vitorsalgado/example/utils/base/ViewPresenter.kt
// package br.com.vitorsalgado.example.android.utils.base
//
// import br.com.vitorsalgado.example.toolkit.Preconditions.checkNotNull
// import io.reactivex.disposables.CompositeDisposable
// import io.reactivex.disposables.Disposable
//
// abstract class ViewPresenter<V : PresenterView> protected constructor(view: V) {
//  private val mCompositeSubscription = CompositeDisposable()
//  protected var view: V? = null
//    private set
//
//  init {
//    checkNotNull(view, "view cannot be null.")
//    this.view = view
//  }
//
//  fun bind(view: V) {
//    checkNotNull(view, "view cannot be null.")
//    this.view = view
//  }
//
//  fun release() {
//    clearSubscriptions()
//    view = null
//  }
//
//  protected fun addSubscription(disposable: Disposable) {
//    mCompositeSubscription.add(disposable)
//  }
//
//  fun clearSubscriptions() {
//    mCompositeSubscription.clear()
//  }
// }
